src/platform/silabs/provision/ProvisionEncoder.h

```diff
@@ -20,24 +20,47 @@
 #include <lib/core/CHIPError.h>
 #include <stddef.h>
 #include <stdint.h>
+#include <cstdlib>
 
 namespace chip {
 namespace DeviceLayer {
 namespace Silabs {
 namespace Provision {
 namespace Encoding {
 
+/*
+begin            out             in               end
+  |---------------v---------------v----------------|
+  |.....offset....|......left.....|.....spare......|
+  |..............size.............|
+  |......................limit.....................|
+*/
+
 struct Buffer
 {
     Buffer(uint8_t * ptr, size_t size, bool at_end = false) { Init(ptr, size, at_end); }
 
     void Init(uint8_t * ptr, size_t size, bool at_end = false)
     {
+        Finish();
+        if(nullptr == ptr)
+        {
+            ptr = static_cast<uint8_t *>(malloc(size));
+            allocated = true;
+        }
         this->begin = ptr;
         this->end   = ptr + size;
         this->in    = at_end ? end : begin;
         this->out   = ptr;
     }
+    void Finish()
+    {
+        if(this->begin && allocated)
+        {
+            free(this->begin);
+        }
+        this->begin = this->end = this->in = this->out = nullptr;
```

Comment:
I'm not sure i follow the changes here - We only seem to free in the first block of the chain. Is there a risk of a memory leak here?